engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedStreamWriter.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.streamprocessor;
+
+import static org.agrona.BitUtil.SIZE_OF_INT;
+import static org.agrona.BitUtil.SIZE_OF_LONG;
+
+import io.camunda.zeebe.logstreams.impl.log.LogEntryDescriptor;
+import io.camunda.zeebe.protocol.Protocol;
+import io.camunda.zeebe.protocol.impl.record.RecordMetadata;
+import io.camunda.zeebe.protocol.record.RecordType;
+import io.camunda.zeebe.protocol.record.RejectionType;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import java.util.function.BinaryOperator;
+import org.agrona.ExpandableDirectByteBuffer;
+import org.agrona.MutableDirectBuffer;
+
+/** Stream writer that writes into a byte buffer instead of writing directly to the stream */
+final class BufferedStreamWriter {
+
+  private static final int INITIAL_BUFFER_CAPACITY = 1024 * 32;
+
+  // todo we can allocate one buffer with max message size here and then it would simplify this -
+  // and limit checking
+  private final MutableDirectBuffer recordBuffer =
+      new ExpandableDirectByteBuffer(INITIAL_BUFFER_CAPACITY);
+
+  private final RecordMetadata metadata = new RecordMetadata();
+
+  private int recordBufferOffset;
+  private int recordCount;
+  private final BinaryOperator<Integer> capacityCalculator;
+
+  BufferedStreamWriter(final BinaryOperator<Integer> capacityCalculator) {
```

Comment:
:+1: TIL about the `BinaryOperator` (though I shouldn't be surprised after the `UnaryOperator` :smile:)